Client/attribute ui

BryonLewis opened the pull request:
This will use the global dataset but set up the initial stage for switching over and making it a smoother transition.
Fixes #437
Old Style:
![oldAttributesInterface](https://user-images.githubusercontent.com/61746913/98673851-f10f3580-2325-11eb-8e84-d61e1a28422a.gif)
New Style:
![NewAttributes](https://user-images.githubusercontent.com/61746913/99295115-fe8b5a80-2812-11eb-8106-a84096a41f64.gif)

Goals:
- Easier to view (List view with clicking on an attribute to edit)
- Easier to stay in this mode while looking at multiple tracks/detections
- Detections can be created/deleted from this panel (Click the settings icon to enter edit mode and delete, modify or create new attributes)
- Streamline attributes for eventual dataset level attribute information stored in the metadata



TODO:  

- [x] Fix Track interface to update style (Fixed so it properly updates when type changed)
- [x] ConfidencePairs needs edit capabilities as well as coloring to quickly indicate types (removing editing for now)
- [x] UI Feedback on the new section
- [x] Look into better separation of data between items (Did what I could do CSS for current layout, need to look into alternate layouts
- [x] Refactor into composition API style for the components instead of old style.
- [x] Need to prevent user from adding a new attribute with the same existing name.
- [x] Suggestion for static widths with truncated hover names for attributes, will attempt to adjust for this
- [ ] tests for the new interface - I think a discussion is needed about how to implement this properly with the different platforms, vuetify and other requirements.
- [ ] Determination of how to handle the old UI (Remove it, leave it?  Leaning towards leaving it for right now)
- [ ] Wiki update for new UI

**Possible Questions and some explanations for feedback:**

- The Track item was modified for a single display tool for the attribute editor at the top, this was to change some behavior it had.
- ConfidencePairs, adding/editing and deleting was left out of this PR, but should be added in a future PR (probably once the sorting is fixed).
- I can break up the AttributesPanel.vue into a subsection component something like `AttributesPanelSection.vue` for tracks/attributes sections because a lot of that is repeated code except for text attribute references and the frame counting.
- I know there is what seems to be unnecessary error handling for the attribute creation/saving/deletion, but this is to inform the user of any eventual permission errors in the metaData versions.  (I.E - a user tries to edit the attributes on someone else's dataset)
- Suggestion on leaving in the `OldAttributesSettings` and items or should we just remove the Settings/Attributes panel from the global settings for now.
- I don't know how you specifically want to do testing for vuetify components when we have platforms that implement vuetify themselves, or if you want to testing at the platform level.  I figured there might need to be a discussion before doing that, but once we have a template for doing it I will create tests for these components.
- Finally do we want to completely remove the older attributes editor/panel or leave it.  We can leave it while we are doing the transition because it still works fine, just with renamed editors to remind myself that eventually it will be removed.

subdavis reviewed (changes requested):
I was working on this yesterday and lost my comments, so here it goes again

## Issues

* Up and down arrows should navigate tracks the same way they do on the other tab, I think.  This is useful as the new review mode UI.
* The track list item checkbox should probably be disabled in the attribute pane.  
* When you move from the attribute panel back to the track list panel, if a track is selected, it isn't scrolled properly in the track list.  If you tap up/down, it's corrected, but it would be nice to have it initialize properly.  This is sorta unrelated to this change, but wanted to mention it.
* setting the value of text and number fields I found a bit frustrating, because you have to hit "enter" for the value to save.  If you hit `escape` or click off the focused input, the value is lost.   IMO, all unfocus events should result in an update.
* Changing the attribute name of an attribute that's already set will result in the value being erased for all uses of that attr. Users will find this frustrating.

## UI Feedback

This is a complicated thing to get right, and I don't know what users are going to like best.  

* I like the ability to edit a single attribute when unused are hidden,
* I'm generally unsure about how this visible/hidden toggle works.  
  * I've noticed that this toggle serves two somewhat conflicting purposes.  1) as a UI optimization, to hide noise and make the attribute panel look cleaner and 2) as a way of exposing the "management" interface where users can manage existing attributes.  I'm concerned that users may be confused by using the same UI for both track/detection-specific editing (setting actual values) and for dataset-general editing of what attributes even exist.  The co-location of these features may lead one to believe that the "options" they set for an attribute apply to a particular track.  It's entirely possible that a more traditional preferences dialog would be better.
  * I think the first purpose is really only a result of vuetify input boxes being so obnoxiously bloated and over-engineered.  Perhaps if all these inputs were replaced with their native form equivalents, they wouldn't need to be hidden for this UI to look clean. I think the track list item is a great example of how removing the vuetify input simplified things.

All that being said, I think this represents an excellent step in the right direction.  The component is really well written and can be refactored in the future if necessary.

I do have a few aesthetic suggestions, which I'll add in a PR.

BryonLewis commented on client/viame-web-common/components/AttributeEditor.vue:
```diff
@@ -0,0 +1,202 @@
+
+<template>
+  <v-dialog
+    v-model="show"
```
I placed it outside of the component and swapped it to use :value.  Don't know why I set it using v-model it was used as value calculated outside of the dialog itself anyways.

BryonLewis commented:
>   * I think the first purpose is really only a result of vuetify input boxes being so obnoxiously bloated and over-engineered.  Perhaps if all these inputs were replaced with their native form equivalents, they wouldn't need to be hidden for this UI to look clean. I think the track list item is a great example of how removing the vuetify input simplified things.

The secondary reason is that you could have a list of disparate things that mean nothing to your specific track.  If there are 20 attributes that can be set and you set index 0,1, 15 to values you then have a whole lot of information that isn't relevant for viewing if you can see all unset data at the same time.  I also didn't want to resort the list of items each time based on what is set.  I think this was a decent middle ground where you have your track level data and you can edit it quickly for set values, but if you want to add more you toggle the visibility (that's why it was a edit icon before).

I do like the native HTML elements more and think they look better.  A bit more annoying to work with (especially the datalist for each text/value input) but it looks cleaner.  The only downside is that the attribute type isn't in there anymore (it used to be the hint for the vuetify input.

![nativeAttributes](https://user-images.githubusercontent.com/61746913/99716729-760cf400-2a76-11eb-85a3-140e0600193e.gif)

subdavis commented:
Oh, nice.  I like that better too.

I think the form type provides enough info about what types are accepted, but that's just me.

BryonLewis commented:
Not ready for review, I want to do the following before it's ready

- [ ] Remove Vuetify section from AttributeInput.vue
- [ ] Refactor AttributeInput.vue into composition component
- [ ] AttributePanel.vue is getting too large, break it up into AttributePanelSection.vue for the Detection/Track attribute sections as well as another section for the Confidence Pairs.  I know it seems like a bit much but I believe it will make it easier to modify in the long run (I think attributes may get more complicated based on some discussions)

BryonLewis commented:
I feel a bit better about this system now.  I think this structure is easier to understand and to modify in the future if needed.

- Swapped inputs all to native HTML elements instead of using vuetify.  (Little annoying with having multiple datalist items, but it works).
- I renamed the main component because it isn't really solely an Attributes Panel.  Renamed it to TracksDetailsPanel.vue.
- Refactored the attributes subsections into a single reusable component with mode `Detection` and `Track`.  Only real different is the `belongsTo` key and the display of the frame numbers.  I pass the core Attributes from the server through the parent object so they don't have to each make a request to the endpoint.
- Added in the small trackScroll fix for the tracklist (just scrolling to the selected on mount) when switching back and forth from TrackDetailsPanel.

subdavis commented on client/viame-web-common/components/AttributeInput.vue:
```diff
@@ -19,58 +24,146 @@ export default Vue.extend({
+    }
+    function onFocus() {
+      if (props.values && props.values.length) {
+        tempVal.value = null;
```
Potentially prefer to just select-all the existing text here.

`inputBoxRef.value.select()` should do it.

BryonLewis replied:
The Datalist thing is annoying in that it won't display the dropdown unless the current value is null for the input element.  So if you have any default `props.values` we need to null it out so you can pick from the list or type in your own.  I'll update it so that if there are no default values (`props.values.length === 0`) it will select the current text to allow typing.  I double checked and this seems to be the same behavior between the type dropdowns in the tracklist and on the MDN site.

subdavis commented on client/viame-web-common/components/AttributesSubsection.vue:
```diff
@@ -0,0 +1,314 @@
+    throw new Error(`Track ${trackId} missing from map`);
+  }
+  return track;
+}
```
should make this a util or something...

subdavis replied:
Maybe it could be hoised from useTrackStore

BryonLewis replied:
I added it into the provides and changed references to where we were using getTrack function or specifically using trackMap for only the purposes of calling getTrack.

subdavis commented on client/viame-web-common/components/TrackDetailsPanel.vue:
```diff
@@ -0,0 +1,394 @@
+  font-size: 0.8em;
+  max-width: 50%;
+  min-width: 50%;
+}
```
unused class

subdavis commented on client/viame-web-common/components/TrackDetailsPanel.vue:
```diff
@@ -0,0 +1,394 @@
+    cursor: pointer;
+    font-weight: bold;
+  }
+}
```
unused class

subdavis commented on client/viame-web-common/components/TrackDetailsPanel.vue:
```diff
@@ -0,0 +1,394 @@
+  font-size: 0.9em;
+  padding: 4px 10px;
+  background-color: #272727;
+}
```
Maybe refactor this to a `BorderHighlight.vue` with a content slot.

Can probably be a pure functional component.

BryonLewis replied:
I added in a `PanelSubsection.vue` component for structuring the data elements that need to be placed in the Panel and have a scrolling subsection as well.  Used for the ConfidenceEditor.vue and both AttributesSubsection.vue components.

I didn't want to go too far into it, but in the future we could add a `settings` slot and perhaps replace the TypeList and TrackList so that everything has a standard style with a header, expandable settings section, and then a scroll area.  It should help with all these vertical subsections that can independently scrolled.

subdavis commented on client/viame-web-common/components/TrackDetailsPanel.vue:
```diff
@@ -0,0 +1,394 @@
+            <span>Add a new Confidence Pair</span>
+          </v-tooltip>
+        </v-row>
+      </div>
```
Refactor to component maybe.

BryonLewis replied:
I figured that was going to be the eventual plan, didn't know if it should be done here or in a new PR.  I can do it here.

subdavis reviewed (changes requested):
Haven't quite finished. Still need to finish understanding TrackDetailsPanel.

Meanwhile, here's some comments.

Also 2 bugs I found best demonstrated by GIF.

## Bug1 

Options disappear in dropdown after they've been selected once

![bug1](https://user-images.githubusercontent.com/4214172/100003406-078ca680-2d94-11eb-8e02-82e4c61b1d34.gif)


## Bug2

Detections don't appear in the "clean" list after they've been saved.  Save button counter goes up, but they're still invisible.

![bug3](https://user-images.githubusercontent.com/4214172/100003558-40c51680-2d94-11eb-881b-6a83831f258e.gif)

BryonLewis commented on client/viame-web-common/components/AttributeInput.vue:
```diff
@@ -19,58 +24,146 @@ export default Vue.extend({
+        switch (newval) {
+          case '':
+          case ' ':
+          case undefined:
```
yup, I added the target && target.value after some refactor afterwards and didn't look closely enough at this.

BryonLewis commented on client/viame-web-common/components/AttributeInput.vue:
```diff
@@ -19,58 +24,146 @@ export default Vue.extend({
-    <v-combobox
+    <datalist
+      v-if="datatype === 'text' && values && values.length"
+      :id="`optionsList_${_uid}`"
```
I'm just adding a comment as a note to make sure that I'm filtering out creating two attributes with the same name.  You can do it technically because the server references stuff by ID but for user confusion and any possible references I do by name I'm going to limit that.  I think that was the reason I initially added the error response to the Editor, just want to confirm that I did.

BryonLewis commented:
Bug 1
Please let me know if I am misinterpreting this bug.

![datalistisbad](https://user-images.githubusercontent.com/61746913/100147271-21e48400-2e69-11eb-8aac-20640bc0bdbb.gif)

Datalist is garbage in it's default behavior.  Once an item is selected it won't display the rest of the options, that is why I need to reset during `onFocus` the value to empty/null.  I believe it's the same reason it is done on the type selection in the `trackItem.vue`.  It does the same thing as well for MDN's demo implementation of the datalist.  It looks like they recognize this and have an example where you can make it work more like it properly should, but it involves creating your own controls for `onClick` as well as manually manipulating the css style to get to behave properly.

I can go through and using `refs`  and `onMount` bind the necessary elements and adjust css position to get it to work like the MDN demo.  I wanted to get feedback before I go about making my own component.

subdavis commented:
>  It does the same thing as well for MDN's demo implementation of the datalist. It looks like they recognize this and have an example where you can make it work more like it properly should, but it involves creating your own controls for onClick as well as manually manipulating the css style to get to behave properly.

Bummer.

I went back and looked again, and this works exactly like the track type dropdown.

I'd say don't worry about it for this PR. The problem is larger in scope than I thought.  Maybe we could log an issue and build our own select component sometime in the future to replace all of them (since track type should work like the mozilla demo too IMO).

Thanks for explaining that.  It isn't ideal for users, but we should maybe take a minute to think if we really want to maintain/support our own `v-select`

subdavis reviewed (changes requested):
It seems the attribute missing bugs from above are still present.

# Issue 1

When changing the selected track with the edit attribute toggled on, the values don't update.

![buga](https://user-images.githubusercontent.com/4214172/100761919-6cae5080-33c1-11eb-98b4-c4a025b5d43a.gif)

# Issue 2

On datasets where the attribute doesn't already exist on at least one track, the attribute won't be visible until it's been saved.

![bugb](https://user-images.githubusercontent.com/4214172/100761920-6d46e700-33c1-11eb-833a-608bc9c28c48.gif)

steps to reproduce:

* Import a new dataset with no annotations
* Create a detection
* Add an attribute
* Show/hide unused.
* Save and reload
* Now it should show up.

BryonLewis commented:
First issue was a bug in the datalist input using a copy of the prop because it needs to be set to nothing when the user clicks on the input so that way the list will properly display.  The tempVal wasn't being modified on new props being set so I added a watcher.

The second issue I feel dumb about.  It was the classic 0 TrackId and not properly checking that the `TrackId !== null` instead of just testing to see if the trackId is truthy.

subdavis reviewed (changes requested):
Couple more small issues.

* Clicking the configure cog only works once.  If you click it, then click anywhere outside the dialog, it won't open anymore.  Reproduce by just spam clicking the cog.  If you use the cancel button inside the dialog, it works properly.

![Screenshot from 2020-12-01 13-40-00](https://user-images.githubusercontent.com/4214172/100782714-cf601600-33da-11eb-950d-f14163517835.png)


* Setting an attribute then immediately clicking the eye makes the attribute disappear on frame 0.
  * Steps to reproduce
    * Click a detection.
    * Open the track attribute settings eye, enter some text
    * Click the eye again
    * Text disaappears, but it's still there if you click the eye back on.
* Detection attributes seems to have the same problem as above on frame 0: detection attributes disappear in when unused are hidden.  This issue is described in a gif in my first review.
* "No track selected" text is shown when a track is selected but has no attributes.  When no track is actually selected, you see "No track attributes set".  These are backward.

BryonLewis commented:
The cog wheel issue was because I wasn't using a v-model for the dialog and was using a value check.  Solved by connecting the `click:outside` event to the `closeEditor` function (This was just a dumb oversight on my part).

The issue with the display not showing up after toggling the visibility was caused by utilizing computed properties and having them update without there being any indication for the computed property to update.  Would require something like watch deep or using refs and manually updating them using functions after an attribute update is called.

There is another option and that is using the similar style to the `TrackListItem.vue` and the track.revision property. 
https://github.com/VIAME/VIAME-Web/blob/master/client/src/components/TrackItem.vue#L58-L80

It looks like it was a created for a similar reason:  Using computed properties derived from a track but the user can edit those tracks and the computed property needs to update properly.  In `TrackListItem` this is when the user clicks on interpolate, delete Feature or add Feature it causes the buttons to refresh by triggering a refresh of the computed function watching the `track.revision` property.  I followed this pattern because the function/watching updating felt messy and why not use the system you already built.

subdavis reviewed (dismissed):
Approved!  Thanks for humoring all my suggestions.

There's just one other issue I noticed:

* Using `escape` key to close the attribute editor dialog has the same bug as clicking off the dialog: you can't open the dialog again until you open the page.
* Also, hitting escape dismisses the dialog _and_ unselects the track, which is kinda jarring because the whole UI changes in an unsuspected way.

I don't know if there's an easy solution to this, or if mousetrap makes it hard.  But it might be nice to fix.  Could also just log an issue and get to it later.

BryonLewis commented:
I'll find see what I can find.  I imagine it has to do with the v-mousetrap and it's global nature.  I know we had that flag used for the global $prompt, may need to make something that prevents it from triggering if dialogs are open.  Hopefully without having to directly interact with `$prompt`  This is something we may run into with other areas so I'll look into fixing it.

BryonLewis commented:
This should prevent the escape from getting to the global mousetrap handler.

subdavis reviewed (changes requested):
Went through a more exhaustive test and found one more thing.

You can't edit attributes anymore.  If you create one, save it, then edit again and try to change its predefined values, you get a message about attribute already exists.  

Also, maybe make the window a tad wider or fix the text wrap?

![Screenshot from 2020-12-03 09-15-39](https://user-images.githubusercontent.com/4214172/101033389-2afbe200-3548-11eb-9852-ade53b88b8ae.png)

BryonLewis commented:
![image](https://user-images.githubusercontent.com/61746913/101084979-c5761880-357c-11eb-9207-c34e8de31184.png)

I welcome any change or suggestion to the warning text that is displayed.  That text is only displayed when the user is editing an existing Attribute.  On brand new attributes that won't be displayed.

I modified the logic so that you can overwrite or change names of the currently selected Attribute.  You aren't allowed to duplicate a name that exists when creating a new attribute or when updating a attribute.  I.E. if you have [foo, bar, buzz], you can't name a new attribute to any of them, but you are allowed to edit the current attribute to a different name as long as it isn't one of the existing items.

subdavis reviewed (approved):
:tada: :tada:  :tada:

BryonLewis commented:
Sorry, I noticed discrepancies in how the Type datalist in `TrackListItem` behaves compared to the Attribute version.  This will make it so hitting enter after clicking will display the old value before setting it.

I also added an issue #477  about fixing whitespace inputs for both the Attributes datalist and the `TrackListItem` Type input.

subdavis commented on client/viame-web-common/components/AttributeInput.vue:
```diff
@@ -50,7 +50,12 @@ export default defineComponent({
-      (e.target as HTMLInputElement).blur();
+      const target = (e.target as HTMLInputElement);
+      // Datalist needs to reset if we blur on no input
+      if (props.values && props.values.length && target.value === '') {
```
target.value will be `' '` not `''` for datalist index 0.  Does this capture that case?

BryonLewis replied:
Selecting that manually from the datalist in the dropdown immediately sets it to undefined and bypasses this function.  This is for only hitting enter after typing, (or specifically not typing anything).  Now hitting enter without adding anything will return it to the default initial value (Same as the type dropdown).  Adding in only spaces will remove the value (same as datalist index 0).

With the current setup if you select the 0 index item it immediately executes the `change`, emitting undefined which will hide the data value if you are editing a single attribute.  So the dropdown bypasses the `blurType` function.

These discrepancies and confusion are why I added that other issue about the behavior of the Type dropdown in `TrackListItem`.

subdavis reviewed (approved):
:+1: :shipit:
